### PR TITLE
PAYARA-3835 Cluster wide data aggregation for Monitoring Console

### PR DESCRIPTION
**Pull request by jbee**
Main changes in this PR:

* adds a module for the core of MC so that it can run within each instance
  * collects the data and does data exchange with the DAS
  * contains core data structures to represent data in memory (including some statistics)
* adds MC webapp module (is manually deployed)
* updates to metric collection due to changes in MP Metrics 2.0

**Dataset Representation**
A bit on the `SeriesDataset` implementation: It uses effectively immutable data to allow consistent snapshots across all series. This mean when adding a point - time + value - to a dataset a new instance is returned representing the change. The returned implementation depends on how the point fits into the points observed so far. If the value is unchanged the `ConstantDataset` is used. If it changes the `PartialDataset` is used. If it seems to become stable again the `StableDataset` is used.

So depending on the value of the series over time we get: 
```
EmptyDataset => ConstantDataset
EmptyDataset => ConstantDataset => PartialDataset
EmptyDataset => ConstantDataset => PartialDataset => StableDataset
EmptyDataset => ConstantDataset => PartialDataset => StableDataset => PartialDataset ...
```

**Data Exchange**
The exchange happens in `InMemoryMonitoringDataRepository` that uses a hazelcast topic to exchange the data. A message `SeriesDatasetsSnapshot` contains all values of the collected series for the collected time. When the message arrives at the listening DAS it is added to the in memory store. Only the DAS stores data. Since members are aware of listening members for a topic the message is only send to the DAS in case it is available. This keeps network traffic as minimal as it can be. The overall network traffic does not increase when length of the in-memory store is enlarged. It increases linear with the number of metric series and instances in the cluster.

### CODE REVIEW

**Comment by jbee** [comment]
jenkins test please

**Comment by jbee** [comment]
@Cousjava that is very unfortunate. Must be related to OSGi or something since I did not change anything AMX related directly in this PR. Have to look into it.

File: appserver/monitoring-console/webapp/src/main/java/fish/payara/monitoring/web/SeriesStatistics.java [hunk 1]
```diff
@@ -0,0 +1,44 @@
+package fish.payara.monitoring.web;
```
Comment (Pandrex247) [inline appserver/monitoring-console/webapp/src/main/java/fish/payara/monitoring/web/SeriesStatistics.java]: Copyright

File: appserver/monitoring-console/webapp/src/main/webapp/index.html [hunk 2]
```diff
@@ -0,0 +1,216 @@
+<!DOCTYPE html>
```
Comment (Pandrex247) [inline appserver/monitoring-console/webapp/src/main/webapp/index.html]: Copyright...?

**Comment by Pandrex247** [comment]
Jenkins seemed to hang trying to complete the quicklook tests - I'll run it again

**Comment by Pandrex247** [comment]
Jenkins test please

**Comment by Pandrex247** [comment]
I also get the NPE:
* Start domain and load Admin Console
* Enable AMX and change monitoring level of all settings to HIGH
* Save config and wait for AMX to try and start (no server restart necessary)

**Comment by Pandrex247** [comment]
It appears to start fine after a server restart, presumably it only throws the NPE when dynamically enabling AMX without a restart

File: appserver/featuresets/payara-web/pom.xml [hunk 3]
```diff
@@ -306,5 +306,13 @@
             <version>${project.version}</version>
             <type>zip</type>
         </dependency>
+
+        <!-- Monitoring Console package -->
+        <dependency>
+            <groupId>org.glassfish.main.packager</groupId>
+            <artifactId>monitoring-console</artifactId>
```
[the remark this answers is not in the document]
Comment (jbee) [reply appserver/featuresets/payara-web/pom.xml]: I'd also think that for now less impact is better. I move it to server.

**Comment by jbee** [comment]
The problem is that `DomainRoot` proxy is `null` - starting to think that the potential for this has always been there but due to some addition this is not triggered reliably. The bean causing this is `amx:pp=,type=domain-root` itself. Maybe the code should not even run for this particular bean?

**Comment by jbee** [comment]
jenkins test please

**Comment by jGauravGupta** [comment]
Tested with latest commits/merge but NPE is still reproducible.
```
  AMX Startup Service: Error loading AMX Beans.
java.lang.RuntimeException: java.lang.NullPointerException
	at org.glassfish.admin.amx.impl.AMXStartupService.loadDomainRoot(AMXStartupService.java:251)
	at org.glassfish.admin.amx.impl.AMXStartupService._loadAMXMBeans(AMXStartupService.java:327)
	at org.glassfish.admin.amx.impl.AMXStartupService.loadAMXMBeans(AMXStartupService.java:224)
.............
	at java.management/com.sun.jmx.mbeanserver.JmxMBeanServer.invoke(JmxMBeanServer.java:801)
	at com.sun.enterprise.v3.admin.DynamicInterceptor.invoke(DynamicInterceptor.java:310)
	at org.glassfish.admin.mbeanserver.BootAMX.bootAMX(BootAMX.java:132)
	at fish.payara.admin.amx.AMXBootService.startup(AMXBootService.java:87)
	at fish.payara.admin.amx.AMXBootService.setEnabled(AMXBootService.java:93)
	at fish.payara.admin.amx.AMXBootService.changed(AMXBootService.java:106)
	at org.jvnet.hk2.config.Transactions$ConfigListenerJob.process(Transactions.java:402)
```

File: appserver/packager/monitoring-console/src/main/assembly/monitoring-console.xml [hunk 4]
```diff
@@ -0,0 +1,64 @@
+<!--
+  DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS HEADER.
+
+  Copyright (c) [2017-2018] Payara Foundation and/or its affiliates.
```
Comment (jGauravGupta) [inline appserver/packager/monitoring-console/src/main/assembly/monitoring-console.xml]: Wrong copyright  year.

File: nucleus/payara-modules/hazelcast-bootstrap/src/main/java/fish/payara/nucleus/store/ClusteredStore.java [hunk 5]
```diff
@@ -38,13 +38,13 @@
  */
```
Comment (jGauravGupta) [inline nucleus/payara-modules/hazelcast-bootstrap/src/main/java/fish/payara/nucleus/store/ClusteredStore.java]: Copyright year should be updated.

File: appserver/monitoring-console/webapp/src/main/java/fish/payara/monitoring/web/MonitoringConsoleResouce.java [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+
+import org.glassfish.internal.api.Globals;
+
+import fish.payara.monitoring.model.Series;
+import fish.payara.monitoring.store.MonitoringDataRepository;
+
+@Path("/")
+@Produces(MediaType.APPLICATION_JSON)
+@RequestScoped
+public class MonitoringConsoleResouce {
```
Comment (jGauravGupta) [inline appserver/monitoring-console/webapp/src/main/java/fish/payara/monitoring/web/MonitoringConsoleResouce.java]: Typo in the class name `MonitoringConsoleResouce `

**Comment by jbee** [comment]
jenkins test please

**Comment by jbee** [comment]
@Pandrex247 , @jGauravGupta , @Cousjava I addressed PR comments. The NPE is some sort of initialisation order issue that is triggered by AMX being also a monitoring data source. Since I did not find a solution I opted to remove AMX as source for now. This is a workaround but AMX as source is not essential and will allow us to test monitoring console in general. AMX source needs to be added in a later release.

**Comment by Pandrex247** [comment]
The issue with quicklook seems to have fixed itself

**Pull request merged**